### ISSUE TITLE
Feature: triple-slash directive to reference declarations that file "implements"

### ISSUE THREAD

**Issue by dawidgarus** [Microsoft/TypeScript#29989]
## Search Terms

implement declaration 

## Suggestion

I would like to introduce new triple-slash directive:
```typescript
/// <reference implements="declarations.d.ts">
```
If you use that directive inside you file it means that it must have the same exports as referenced declaration. In other words - if you created d.ts file out of it, it would be the same as referenced declaration.

## Use Cases

Consider example from docs: https://www.typescriptlang.org/docs/handbook/module-resolution.html.
You have virtual directories:
```json
{
  "compilerOptions": {
    "rootDirs": [
      "src/en",
      "src/pl",
      "src/$locale"
    ]
  }
}
```
src/$locale/messages.d.ts:
```typescript
export declare const messages: {
  hello: string;
  title: string;
};
```
src/en/messages.ts:
```typescript
/// <reference implements="../$locale/messages.d.ts">
export const messages = {
  hello: 'Hello',
  title: 'Homepage',
};
```
src/pl/messages.ts:
```typescript
/// <reference implements="/../$locale/messages.d.ts">
export const messages = {
  hello: 'Witaj',
  title: 'Strona domowa',
};
```
The directive would tell the compiler to check whether the file actually "implements" the declaration and it should produce error if not:
```typescript
/// <reference implements="../$locale/messages.d.ts">
export const messages = {
  helo: 'Hello', // ERROR: {helo, title} is not compatible with {hello, title}
  title: 'Homepage',
};
```

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by RyanCavanaugh** [comment]
Duplicate #420
